snuba/request_schema.py

```diff
@@ -0,0 +1,132 @@
+            'type': 'object',
+            'properties': {},
+            'required': [],
+            'definitions': {},
+            'additionalProperties': False,
+        }
+
+        for schema in itertools.chain([self.__query_schema], self.__extension_schemas.values()):
+            assert schema['type'] == 'object', 'subschema must be object'
+            assert schema['additionalProperties'] is False, 'subschema must not allow additional properties'
+            self.__composite_schema['required'].extend(schema.get('required', []))
+
+            for property_name, property_schema in schema['properties'].items():
+                assert property_name not in self.__composite_schema['properties'], 'subschema cannot redefine property'
+                self.__composite_schema['properties'][property_name] = property_schema
+
+            for definition_name, definition_schema in schema.get('definitions', {}).items():
+                assert definition_name not in self.__composite_schema['definitions'], 'subschema cannot redefine definition'
+                self.__composite_schema['definitions'][definition_name] = definition_schema
+
+        self.__composite_schema['required'] = set(self.__composite_schema['required'])
+
+    @classmethod
+    def build_with_extensions(cls, extensions: Mapping[str, QueryExtension]) -> RequestSchema:
+        generic_schema = GENERIC_QUERY_SCHEMA
+        extensions_schemas = {
+            extension_key: extension.get_schema()
+            for extension_key, extension
+            in extensions.items()
+        }
+        return cls(generic_schema, extensions_schemas)
+
+    def validate(self, value) -> Request:
+        value = validate_jsonschema(value, self.__composite_schema)
+
+        query_body = {key: value.pop(key) for key in self.__query_schema['properties'].keys() if key in value}
+
+        extensions = {}
+        for extension_name, extension_schema in self.__extension_schemas.items():
+            extensions[extension_name] = {key: value.pop(key) for key in extension_schema['properties'].keys() if key in value}
+
+        return Request(Query(query_body), extensions)
+
+    def __generate_template_impl(self, schema) -> Any:
+        """
+        Generate a (not necessarily valid) object that can be used as a template
+        from the provided schema
+        """
+        typ = schema.get('type')
+        if 'default' in schema:
+            default = schema['default']
+            return default() if callable(default) else default
+        elif typ == 'object':
+            return {prop: self.__generate_template_impl(subschema) for prop, subschema in schema.get('properties', {}).items()}
+        elif typ == 'array':
+            return []
+        elif typ == 'string':
+            return ""
+        return None
+
+    def generate_template(self) -> Any:
+        return self.__generate_template_impl(self.__composite_schema)
+
+
+def validate_jsonschema(value, schema, set_defaults=True):
```

Comment:
I think this should actually stay in `snuba.schemas`, since it's not really tied to the request.